### PR TITLE
[inductor] Added decomposition for _upsample_bilinear2d_aa

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Added decomposition for _upsample_bilinear2d_aa

Benchmark results:
- CPU decomposition is very slow vs Eager
  - about 3 times slower with a "single" kernel vs multiple kernels.
- CUDA perfs are comparable on large inputs. On small inputs "Compiled" column results are bounded by a min exec time ~60us (used infrastructure) and "Just Triton" column shows the time taken by executing triton compiled code only without dynamo etc (See source below for benchcode details).
```
[-------------------------------------------- Interpolate bilinear, AA=true, cpu -------------------------------------------]
                                                                                      |   Eager   |  Compiled  |  Just C++
1 threads: ------------------------------------------------------------------------------------------------------------------
      Input (1, 3, 500, 400) -> (256, 256), torch.uint8, torch.contiguous_format      |    766.0  |   20620.2  |     20677.3
      Input (1, 3, 500, 400) -> (256, 256), torch.float32, torch.contiguous_format    |   1719.6  |   19530.7  |     19497.4
      Input (1, 3, 500, 400) -> (256, 256), torch.uint8, torch.channels_last          |    333.1  |   20856.8  |     21364.1
      Input (1, 3, 500, 400) -> (256, 256), torch.float32, torch.channels_last        |   2199.4  |   19880.4  |     19793.3
      Input (4, 3, 500, 400) -> (256, 256), torch.uint8, torch.contiguous_format      |   3217.5  |   83345.9  |     83517.6
      Input (4, 3, 500, 400) -> (256, 256), torch.float32, torch.contiguous_format    |   7111.5  |   75855.2  |     75608.2
      Input (4, 3, 500, 400) -> (256, 256), torch.uint8, torch.channels_last          |   1300.4  |   83555.3  |     83619.2
      Input (4, 3, 500, 400) -> (256, 256), torch.float32, torch.channels_last        |   8881.5  |   86082.9  |     85837.3

      Input (1, 3, 1200, 1300) -> (200, 300), torch.uint8, torch.contiguous_format    |   3577.0  |   15391.7  |     15341.8
      Input (1, 3, 1200, 1300) -> (200, 300), torch.float32, torch.contiguous_format  |   9438.9  |   12377.3  |     12249.9
      Input (1, 3, 1200, 1300) -> (200, 300), torch.uint8, torch.channels_last        |   1130.9  |   12413.4  |     12326.8
      Input (1, 3, 1200, 1300) -> (200, 300), torch.float32, torch.channels_last      |  10247.6  |    8201.1  |      8060.1
      Input (4, 3, 1200, 1300) -> (200, 300), torch.uint8, torch.contiguous_format    |  13913.0  |   62827.1  |     62963.2
      Input (4, 3, 1200, 1300) -> (200, 300), torch.float32, torch.contiguous_format  |  37462.7  |   48888.1  |     48815.6
      Input (4, 3, 1200, 1300) -> (200, 300), torch.uint8, torch.channels_last        |   4723.9  |   51666.5  |     51516.6
      Input (4, 3, 1200, 1300) -> (200, 300), torch.float32, torch.channels_last      |  41864.6  |   34116.5  |     33966.8

      Input (1, 3, 300, 400) -> (600, 700), torch.uint8, torch.contiguous_format      |   1864.5  |   35455.5  |     35566.7
      Input (1, 3, 300, 400) -> (600, 700), torch.float32, torch.contiguous_format    |   2653.1  |   35303.6  |     35455.9
      Input (1, 3, 300, 400) -> (600, 700), torch.uint8, torch.channels_last          |    510.8  |   35767.9  |     35901.2
      Input (1, 3, 300, 400) -> (600, 700), torch.float32, torch.channels_last        |   5365.5  |   33153.8  |     33548.8
      Input (4, 3, 300, 400) -> (600, 700), torch.uint8, torch.contiguous_format      |   8824.6  |  141773.7  |    139218.2
      Input (4, 3, 300, 400) -> (600, 700), torch.float32, torch.contiguous_format    |  10564.3  |  140867.2  |    140051.1
      Input (4, 3, 300, 400) -> (600, 700), torch.uint8, torch.channels_last          |   2093.3  |  153663.1  |    154263.3
      Input (4, 3, 300, 400) -> (600, 700), torch.float32, torch.channels_last        |  21512.9  |  132679.4  |    132839.2

Times are in microseconds (us).

[------------------------------------------- Interpolate bilinear, AA=true, cuda --------------------------------------------]
                                                                                        |  Eager   |  Compiled  |  Just Triton
1 threads: -------------------------------------------------------------------------------------------------------------------
      Input (1, 3, 500, 400) -> (256, 256), torch.float32, torch.contiguous_format      |    11.3  |     65.3   |       22.3
      Input (1, 3, 500, 400) -> (256, 256), torch.float32, torch.channels_last          |    29.3  |     70.5   |       23.8
      Input (4, 3, 500, 400) -> (256, 256), torch.float32, torch.contiguous_format      |    42.1  |     70.1   |       33.8
      Input (4, 3, 500, 400) -> (256, 256), torch.float32, torch.channels_last          |    91.8  |     64.9   |       47.9

      Input (1, 3, 1200, 1300) -> (200, 300), torch.float32, torch.contiguous_format    |    85.9  |    117.2   |      117.9
      Input (1, 3, 1200, 1300) -> (200, 300), torch.float32, torch.channels_last        |   216.9  |    163.9   |      166.0
      Input (4, 3, 1200, 1300) -> (200, 300), torch.float32, torch.contiguous_format    |   331.8  |    471.0   |      472.0
      Input (4, 3, 1200, 1300) -> (200, 300), torch.float32, torch.channels_last        |   841.7  |    647.4   |      650.8

      Input (1, 3, 300, 400) -> (600, 700), torch.float32, torch.contiguous_format      |    30.1  |     64.7   |       28.4
      Input (1, 3, 300, 400) -> (600, 700), torch.float32, torch.channels_last          |    54.2  |     65.7   |       39.3
      Input (4, 3, 300, 400) -> (600, 700), torch.float32, torch.contiguous_format      |    95.6  |    109.8   |      109.5
      Input (4, 3, 300, 400) -> (600, 700), torch.float32, torch.channels_last          |   187.8  |    232.8   |      232.8

      Input (1, 3, 2345, 2456) -> (1234, 1345), torch.float32, torch.contiguous_format  |   246.0  |    225.0   |      229.2
      Input (1, 3, 2345, 2456) -> (1234, 1345), torch.float32, torch.channels_last      |   778.7  |    575.0   |      574.9
      Input (4, 3, 2345, 2456) -> (1234, 1345), torch.float32, torch.contiguous_format  |   872.5  |    898.3   |      898.7
      Input (4, 3, 2345, 2456) -> (1234, 1345), torch.float32, torch.channels_last      |  2996.3  |   2289.4   |     2288.8

      Input (1, 3, 1234, 1345) -> (2345, 2456), torch.float32, torch.contiguous_format  |   431.3  |    381.1   |      381.2
      Input (1, 3, 1234, 1345) -> (2345, 2456), torch.float32, torch.channels_last      |   796.4  |   1042.8   |     1044.2
      Input (4, 3, 1234, 1345) -> (2345, 2456), torch.float32, torch.contiguous_format  |  1323.6  |   1518.6   |     1515.9
      Input (4, 3, 1234, 1345) -> (2345, 2456), torch.float32, torch.channels_last      |  2797.8  |   3728.6   |     3728.0

      Input (1, 3, 2345, 2456) -> (120, 200), torch.float32, torch.contiguous_format    |   461.7  |    435.9   |      436.3
      Input (1, 3, 2345, 2456) -> (120, 200), torch.float32, torch.channels_last        |   941.6  |    521.6   |      521.3
      Input (4, 3, 2345, 2456) -> (120, 200), torch.float32, torch.contiguous_format    |  1832.3  |   1724.6   |     1725.6
      Input (4, 3, 2345, 2456) -> (120, 200), torch.float32, torch.channels_last        |  3740.3  |   2040.9   |     2041.0

Times are in microseconds (us).
```
[Source](https://github.com/vfdev-5/pth-inductor-dev/blob/862599e5cbece5c41068cd16cc3afbf25f26be39/perf_interp_bilinear_aa_vs_gencode.py)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @Xia-Weiwen @aakhundov

### CODE REVIEW

**Review by Chillee — APPROVED** [review]
So we don't turn this on in Inductor because the benchmark numbers are worse than eager, right?

**Comment by Chillee** [comment]
If I'm understanding the benchmarks correctly, it seems like we shouldn't enable it right now? Since even for fp32 on cuda, perf is quite a bit worse than the eager impl.

**Comment by vfdev-5** [comment]
yes, you are right, we should not enable it right now. Let me see how to do that in a minimal way

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -3142,13 +3143,136 @@ def gru_impl(
...
+        assert interp_dim % 4 == 2
+        in_x = torch.arange(in_w, device=in_tensor.device).view((1, 1, 1, in_w))
+        src_idx_min = src_idx_min.view(1, 1, out_size, 1)
+
+        in_tensor_list = [
+            in_tensor[
+                n_idx, c_idx, torch.clamp(src_idx_min + k, max=in_h - 1), in_x
+            ].contiguous(memory_format=memory_format)
+            for k in range(max_interp_size)
+        ]
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:3241]: This is *surely* causing a massive perf hit. The way to do this is via some broadcasting over the indices

File: torch/_inductor/triton_heuristics.py [hunk 2]
```diff
@@ -1114,6 +1114,16 @@ def persistent_reduction(size_hints, reduction_hint=False, meta=None, filename=N
...
+    # TODO: Check if this can be helpful in general.
+    # https://github.com/pytorch/pytorch/pull/111656
+    # This brings a speed-up with interpolate bilinear aa=True op on batched input
+    elif reduction_hint == ReductionHint.DEFAULT:
+        configs += [
+            triton_config_reduction(size_hints, xblock, rnumel)
+            for xblock in (256, 512)
+            if rnumel * xblock <= 4096 * 4 and xblock <= xnumel
+        ]
+
```
Comment (lezcano) [inline torch/_inductor/triton_heuristics.py:1126]: This has to go in a separate PR and benchmarked on its own.
It'd be good to benchmark just adding 512 first, and then perhaps adding both, as this may increase compilation times of the models.

**Review by lezcano — CHANGES_REQUESTED** [review]

File: torch/_inductor/triton_heuristics.py (hunk 2 above)
Comment (vfdev-5) [reply torch/_inductor/triton_heuristics.py:1126]: Separate PR with adding 256 and 512 only is here: https://github.com/pytorch/pytorch/pull/111656

**Comment by lezcano** [comment]
@vfdev-5 do we really want to close this one?
